packages/ai-jsx/src/lib/openai.tsx

```diff
@@ -341,8 +347,60 @@ export async function* OpenAIChatModel(
     promptTokenLimit
   );
 
-  const messages: OpenAIClient.Chat.CreateChatCompletionRequestMessage[] = await Promise.all(
-    conversationMessages.map(async (message) => {
+  // OpenAI requires that tool calls with IDs be immediately followed by tool messages with the corresponding IDs.
+  // Walk through the conversation to find IDs that don't match up. Anything that doesn't match will be represented
+  // as ID-less function calls.
+  let activeFunctionCalls: string[] = [];
+  let activeFunctionResponses: string[] = [];
+  const unmatchedIds = new Set<string>(); // IDs that should be represented as ID-less function calls because they were part of an unmatched block.
+
+  const flushActiveIds = () => {
+    const dedupedFunctionCalls = new Set(activeFunctionCalls);
+    const dedupedFunctionResponses = new Set(activeFunctionResponses);
+
+    // If there were any duplicated IDs, or any IDs were in one set but not the other, _all_ the IDs are invalid.
+    const isInvalid =
+      dedupedFunctionCalls.size !== activeFunctionCalls.length ||
+      dedupedFunctionResponses.size !== activeFunctionResponses.length ||
+      !activeFunctionCalls.every((id) => dedupedFunctionResponses.has(id)) ||
+      !activeFunctionResponses.every((id) => dedupedFunctionCalls.has(id));
+
+    if (isInvalid) {
+      for (const id of activeFunctionCalls) {
+        unmatchedIds.add(id);
+      }
+      for (const id of activeFunctionResponses) {
+        unmatchedIds.add(id);
+      }
+    }
```

Comment:
Isn't this a bit harsh? Shouldn't we just set the ones that didn't match to `unmatchedIds` instead of invalidating the whole run?

Reply:
The challenge is that the API asserts the invariant very strictly -- so the entire batch needs to fully matched/valid or else OAI will reject the request.

But it does allow a single request to have mixed old- and new-style messages -- how the response is emitted is determined by whether the functions were _defined_ as tools vs. functions.